exporter/tanzuobservabilityexporter/metrics_test.go

```diff
@@ -197,6 +196,270 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+	metric := newHistogramMetricWithDataPoints(
+		"bad.metric", pdata.MetricAggregationTemporalityUnspecified, nil)
+	sender := &mockGaugeSender{}
+	observedZapCore, observedLogs := observer.New(zap.DebugLevel)
+	consumer := newHistogramConsumer(
+		&mockHistogramDataPointConsumer{},
+		nil,
+		sender,
+		&consumerOptions{
+			Logger:                zap.New(observedZapCore),
+			ReportInternalMetrics: true,
+		},
+	)
+	assert.Equal(t, pdata.MetricDataTypeHistogram, consumer.Type())
+	var errs []error
+	expectedNoAggregationCount := 3
+	for i := 0; i < expectedNoAggregationCount; i++ {
+		consumer.Consume(metric, &errs)
+	}
+	consumer.PushInternalMetrics(&errs)
+
+	assert.Len(t, errs, 0)
+	assert.Contains(t, sender.metrics, tobsMetric{
+		Name:  noAggregationTemporalityMetricName,
+		Value: float64(expectedNoAggregationCount),
+		Tags:  map[string]string{"type": "histogram"},
+	})
+	allLogs := observedLogs.All()
+	assert.Len(t, allLogs, expectedNoAggregationCount)
+}
+
+func TestHistogramReporting(t *testing.T) {
+	observedZapCore, observedLogs := observer.New(zap.DebugLevel)
+	report := newHistogramReporting(zap.New(observedZapCore))
+	metric := newMetric("a.metric", pdata.MetricDataTypeHistogram)
+	leTagInUseCount := 2
+	for i := 0; i < leTagInUseCount; i++ {
+		report.LogLeTagInUse(metric)
+	}
+	malformedCount := 3
+	for i := 0; i < malformedCount; i++ {
+		report.LogMalformed(metric)
+	}
+	noAggregationTemporalityCount := 5
+	for i := 0; i < noAggregationTemporalityCount; i++ {
+		report.LogNoAggregationTemporality(metric)
+	}
+
+	assert.Equal(t, int64(leTagInUseCount), report.LeTagInUse())
+	assert.Equal(t, int64(malformedCount), report.Malformed())
+	assert.Equal(t, int64(noAggregationTemporalityCount), report.NoAggregationTemporality())
+	assert.Equal(
+		t,
+		leTagInUseCount+malformedCount+noAggregationTemporalityCount,
+		observedLogs.Len())
+
+	sender := &mockGaugeSender{}
+	var errs []error
+
+	report.Report(sender, &errs)
+
+	assert.Empty(t, errs)
+	assert.Contains(
+		t,
+		sender.metrics,
+		tobsMetric{
+			Name:  leInUseMetricName,
+			Value: float64(leTagInUseCount),
+		})
+	assert.Contains(
+		t,
+		sender.metrics,
+		tobsMetric{
+			Name:  malformedHistogramMetricName,
+			Value: float64(malformedCount),
+		})
+	assert.Contains(
+		t,
+		sender.metrics,
+		tobsMetric{
+			Name:  noAggregationTemporalityMetricName,
+			Value: float64(noAggregationTemporalityCount),
+			Tags:  typeIsHistogramTags,
+		})
+}
+
+func TestHistogramReportingError(t *testing.T) {
+	report := newHistogramReporting(zap.NewNop())
+	sender := &mockGaugeSender{errorOnSend: true}
+	var errs []error
+
+	report.Report(sender, &errs)
+
+	assert.NotEmpty(t, errs)
+}
+
+func TestCumulativeHistogramDataPointConsumer(t *testing.T) {
+	metric := newMetric("a.metric", pdata.MetricDataTypeHistogram)
+	histogramDataPoint := pdata.NewHistogramDataPoint()
+
+	// Creates bounds of -Inf to <=2.0; >2.0 to <=5.0; >5.0 to <=10.0; >10.0 to +Inf
```

Comment:
really like having this comment explicitly spell out the bounds :-)

Reply:
Thank you.